File: libs/east-diagnostics/src/rules/no-let-const-in-expression.ts
/**
 * Copyright (c) 2025 Elara AI Pty Ltd
 * Dual-licensed under AGPL-3.0 and commercial license. See LICENSE for details.
 */
import type * as ts from "typescript";
import type { EastRule } from "../types.js";
import { matchBlockBuilderCall } from "../block-builder.js";

const NAME = "no-let-const-in-expression";
const CODE = 990008;

// `$.let` / `$.const` declare a variable in the block (a statement-like effect)
// and return the expression handle. The only correct position is the initializer
// of a JS `const`/`let`. Used anywhere else — as an argument (`$.if($.let(...))`),
// a chain target (`$.let(...).add(1n)`), or an operand — it buries a declaration
// inside an expression. Type-checks fine, so the checker won't catch it.
export const noLetConstInExpression: EastRule = {
  name: NAME,
  code: CODE,
  description:
    "Require $.let/$.const to be bound to a const; disallow using the result inline in an expression.",
  check(node, ctx) {
    const match = matchBlockBuilderCall(node, ctx);
    if (match === undefined) return;

    const t = ctx.ts;
    const call = match.call;

    // Walk out of wrapping parentheses.
    let current: ts.Node = call;
    let parent = current.parent;
    while (parent !== undefined && t.isParenthesizedExpression(parent)) {
      current = parent;
      parent = parent.parent;
    }
    if (parent === undefined) return;

    // Allow-list: `$.let`/`$.const` is only valid as a JS `const`/`let`
    // initializer, a bare statement, a `return`, or a concise arrow body. ANY
    // other position buries the declaration in an expression — a struct-field
    // value (`field: $.let(...)`), an array element, a call argument
    // (`$.if($.let(...))`), a chain target (`$.let(...).add(1n)`), an operator
    // operand, a ternary branch, etc. Flag everything that is not allowed.
    const allowed =
      (t.isVariableDeclaration(parent) && parent.initializer === current) ||
      (t.isExpressionStatement(parent) && parent.expression === current) ||
      (t.isReturnStatement(parent) && parent.expression === current) ||
      (t.isArrowFunction(parent) && parent.body === current);
    if (allowed) return;

    const sf = ctx.sourceFile;
    const start = call.getStart(sf);
    ctx.report({
      ruleName: NAME,
      code: CODE,
      start,
      length: call.getEnd() - start,
      messageText: `\`$.${match.method}\` declares a variable — bind it to a \`const\` first (\`const x = $.${match.method}(value, Type)\`), don't use the result inline (e.g. \`$.if($.let(...))\` or \`$.let(...).add(...)\`).`,
      category: "warning",
    });
  },
};

File: libs/east-py/packages/east-py/east/diagnostics/rules/no_let_const_in_expression.py
#
# Copyright (c) 2025 Elara AI Pty Ltd
# Licensed under the Business Source License 1.1. See LICENSE.md for details.
#
"""``no-let-const-in-expression``: ``b.let`` / ``b.const`` declare a binding
in the block and hand back its handle. The only readable position for that is
a statement of its own. Buried in an expression — a call argument, a struct
field's value, an element, the target of a chain — it hides a declaration
inside something that reads as a value, and it type-checks fine. The
TypeScript rule of the same name.
"""

from __future__ import annotations

import ast

from east.diagnostics.types import Body, Context, body_nodes


def message(method: str) -> str:
    return (f"b.{method}(...) declares a binding — give it its own statement "
            f"(`x = b.{method}(value, Type)`) and use `x`, rather than burying the "
            "declaration inside an expression")


class NoLetConstInExpression:
    name = "no-let-const-in-expression"
    code = 13
    category = "warning"
    supersedes: tuple[str, ...] = ()
    description = "b.let / b.const belongs on its own statement, never buried inside an expression."

    def check(self, body: Body, ctx: Context) -> None:
        if body.block is None:
            return
        parents = _parents(body)
        for node in body_nodes(body):
            if not (isinstance(node, ast.Call) and isinstance(node.func, ast.Attribute)
                    and node.func.attr in ("let", "const")
                    and isinstance(node.func.value, ast.Name) and node.func.value.id == body.block):
                continue
            parent = parents.get(id(node))
            if parent is None or _is_binding_position(parent, node):
                continue
            ctx.report(node, self, message(node.func.attr))


def _is_binding_position(parent: ast.AST, node: ast.AST) -> bool:
    """The positions where a declaration reads as a declaration: the value of
    an assignment, a bare statement, a ``return``, or a lambda's whole body."""
    if isinstance(parent, (ast.Assign, ast.AnnAssign, ast.AugAssign)) and parent.value is node:
        return True
    if isinstance(parent, ast.Expr) and parent.value is node:
        return True
    if isinstance(parent, ast.Return) and parent.value is node:
        return True
    return isinstance(parent, ast.Lambda) and parent.body is node


def _parents(body: Body) -> dict[int, ast.AST]:
    parents: dict[int, ast.AST] = {}
    for node in body_nodes(body):
        for child in ast.iter_child_nodes(node):
            parents[id(child)] = node
    return parents
